### PR TITLE
Fix slang block inline

### LINKED ISSUE #430: Fix Slang block inlining
This method 
https://github.com/pharo-project/pharo-vm/blob/81891cf884f8b42e694bc178f63239df2aae43f8/smalltalksrc/VMMaker/InterpreterPrimitives.class.st#L5039-L5052
cannot be translated.

But this one yes:
https://github.com/pharo-project/pharo-vm/blob/81891cf884f8b42e694bc178f63239df2aae43f8/smalltalksrc/VMMaker/InterpreterPrimitives.class.st#L4999-L5010

### PR DESCRIPTION
**Pull request by guillep**
Fixes: https://github.com/pharo-project/pharo-vm/issues/430
Add proper support for

- return blocks with assignments
- assing blocks with returns
- return blocks with returns
- assign blocks with assignments

Remove duplicated code and "hacks"

### CODE REVIEW

File: smalltalksrc/Slang/TAssignmentNode.class.st
Comment (guillep) [inline smalltalksrc/Slang/TAssignmentNode.class.st]: Careful with this before merge

Comment (PalumboN) [reply smalltalksrc/Slang/TAssignmentNode.class.st]: Why "careful"? We can delete it, right?

```suggestion
```

Comment (guillep) [reply smalltalksrc/Slang/TAssignmentNode.class.st]: By careful I meant "let's wait until CI runs ok!"

Comment (guillep) [reply smalltalksrc/Slang/TAssignmentNode.class.st]: @PalumboN this is failing when generating the JIT VM :)

File: extracted/vm/include/common/sqMemoryAccess.h
Comment (PalumboN) [inline extracted/vm/include/common/sqMemoryAccess.h]: ⚠️  Missing parameter in `uint64Atput`? Should be as `long64Atput`, right?

File: smalltalksrc/Melchor/ManifestMelchor.class.st
Comment (PalumboN) [inline smalltalksrc/Melchor/ManifestMelchor.class.st]: ????

File: smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st
Comment (PalumboN) [inline smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st:156]: These `withoutParentheses` were added because the expressions have parentheses by default? Was not better to change the default? 🤔

**Review by PalumboN — APPROVED** [review]

File: smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st
Comment (guillep) [reply smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st:156]: Actually, expressions have no parentheses by default, but assignments do.
What I did here is to decide if a child needs parentheses in the parent node, which is better than a global strategy.
However, the precedence in C is complex (https://en.cppreference.com/w/c/language/operator_precedence) so:
 - the simplest is to parenthesize everything. Uglier but correct code
 - the hardest is to embed C precedence into the pretty printer and put parentheses only if necessary (i.e., if the parent has less precedence than the child in the tree)

File: extracted/vm/include/common/sqMemoryAccess.h
Comment (guillep) [reply extracted/vm/include/common/sqMemoryAccess.h]: yes

**Comment by privat** [comment]
This PR is really hard to review :(

**Comment by guillep** [comment]
Yes, i agree with that. I worked yesterday on it to evaluate how far it was from success. I think I tried doing too much. I will close this PR and start a couple of new ones using the code here as basis.

I will split assignment from returns, and probably add it as a canonicalisation pass in the tree after inlining.What I saw is that canonicalising the tree (so there are no returns or assignments of statements) has an impact on inlining decisions. So it makes it a bit more difficult to debug if it is done before inlining.Also I observed many cases where this canonicalisation need to do rewrites in the tree. In particular,

```
return a ifTrue: [ x]
```

Should introduce an else branch to be semantically equivalent.

```
a ifTrue: [ return x] ifFalse: [ return null]
```

One way to reduce side effects is to only canonicalise in this way only when strictly required (i.e. If the content of one of the block arguments is not a valid a expression, e.g., a return statement).And this should manage also ifNil: with its pair ifNotNil: that receives the receiver expression as argument, and switch/case statements, making it less trivial. :)

Oh yes, i kind of convinced myself that restarting and targetting a smaller change is the way to go

**Pull request closed without merging**
